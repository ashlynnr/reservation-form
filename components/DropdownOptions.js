import React from "react";
import styled from "styled-components";

const DropdownOptions = ({
  title,
  ages,
  currentValue,
  handleChange,
  guestType,
  checked,
  options
}) => {
  return (
    <Section>
      {title} <br />
      {ages}
      <Dropdown
        value={currentValue ? currentValue : 1}
        onChange={e => handleChange(e, guestType)}
      >
        {options.map((guests, index) => (
          <option key={guests + index} value={guests} disabled={!checked}>
            {guests}
          </option>
        ))}
      </Dropdown>
    </Section>
  );
};

export default DropdownOptions;

const Section = styled.div`
  font-size: 1.1em;
  display: grid;
  grid-gap: 5px;
  align-items: center;
`;
const Dropdown = styled.select`
  width: 50px;
  font-size: 1.05em;
`;
