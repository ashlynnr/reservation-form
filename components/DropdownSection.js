import React from "react";
import styled from "styled-components";
import { orderHandler } from "../utils/helpers";
import DropdownOptions from "./DropdownOptions";

const DropdownSection = ({
  adults,
  children,
  checked,
  id,
  reviseOrder,
  order
}) => {
  const handleChange = (e, key) => {
    const value = parseInt(e.target.value);
    orderHandler({ id, value, key, reviseOrder, order });
  };

  const currentValueAdults = checked
    ? order.filter(reservation => reservation.id === id)[0].adults
    : null;

  const currentValueChildren = checked
    ? order.filter(reservation => reservation.id === id)[0].children
    : null;

  return (
    <Container>
      <DropdownOptions
        title={"Adults"}
        ages={"(18+)"}
        currentValue={currentValueAdults}
        handleChange={handleChange}
        guestType={"adults"}
        checked={checked}
        options={adults}
      />
      <DropdownOptions
        title={"Children"}
        ages={"(0-17)"}
        currentValue={currentValueChildren}
        handleChange={handleChange}
        guestType={"children"}
        checked={checked}
        options={children}
      />
    </Container>
  );
};

export default DropdownSection;

const Container = styled.div`
  padding: 20px 10px 40px;
  display: grid;
  grid-template-columns: 50% 50%;
  height: 100%;
  justify-items: center;
  align-items: center;
`;
