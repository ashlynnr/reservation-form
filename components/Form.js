import React, { useEffect } from "react";
import styled from "styled-components";
import DropdownOptions from "./DropdownOptions";
import { editReservation } from "../utils/helpers";

const Form = ({ room, order, reviseOrder, index }) => {
  const checkHandler = e => {
    if (!e.target.checked) {
      removeRooms();
    } else {
      addRooms();
    }
  };

  const orderHandler = (id, value, key) => {
    const updatedOrder = order.map(a => Object.assign({}, a));

    updatedOrder.map((data, i) => {
      if (data.id === id) {
        data[key] = value;
      }
    });
    reviseOrder(updatedOrder);
  };

  const removeRooms = () => {
    const updatedOrder = order.map(a => Object.assign({}, a));
    let removals = 1;
    updatedOrder.map(data => {
      if (data.id > room.id) {
        removals++;
      }
    });
    updatedOrder.splice(index, index + removals);
    reviseOrder(updatedOrder);
  };

  const addRooms = () => {
    const updatedOrder = order.map(a => Object.assign({}, a));
    let firstRoom = 1;
    const lastRoom = room.id;
    let additions = [];

    while (firstRoom <= lastRoom) {
      additions.push(firstRoom++);
    }

    additions.map(roomId => {
      if (!order.some(data => data.id === roomId)) {
        updatedOrder.push({ id: roomId, adults: 1, children: 0 });
      }
    });
    reviseOrder(updatedOrder);
  };

  const checked = order.some(data => data.id === room.id);

  return (
    <Container checked={checked}>
      <Header>
        {room.optional && (
          <Checkbox
            type="checkbox"
            name="room"
            value={room.name}
            onClick={checkHandler}
            checked={checked}
          />
        )}
        {room.name}
      </Header>
      <Section checked={checked}>
        <DropdownOptions
          adults={room.adults}
          children={room.children}
          checked={checked}
          room={room}
          orderHandler={orderHandler}
          order={order}
        />
      </Section>
    </Container>
  );
};

export default Form;

const Container = styled.div`
  display: grid;
  grid-template-rows: 25% 75%;
  height: 100%;
  width: 100%;
  background: ${props =>
    props.checked ? props.theme.grey : props.theme.disabledLightPurple};
  border: ${props =>
    !props.checked
      ? `solid ${props.theme.disabledPurple} 5px`
      : `solid ${props.theme.grey} 5px`};
  border-radius: 2%;
  transition: 0.3s all ease-in;
`;

const Section = styled.div`
  height: 98%;
  background: ${props => (props.checked ? "white" : "transparent")};
  border-radius: 2%;
  grid-template-columns: 2fr;
  transition: 0.3s background ease-in;
`;

const Header = styled.div`
  font-weight: bold;
  font-size: 1.2em;
  height: 100%;
  display: flex;
  align-items: center;
  padding: 0 5px;
`;

const Checkbox = styled.input`
  -ms-transform: scale(1.5);
  -moz-transform: scale(1.5);
  -webkit-transform: scale(1.5);
  -o-transform: scale(1.5);
  transform: scale(1.5);
  margin-right: 10px;
`;
