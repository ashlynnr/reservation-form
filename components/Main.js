import React, { useState } from "react";
import styled from "styled-components";
import rooms from "../utils/data";
import Form from "./Form";

const Main = () => {
  const [counter, setCounter] = useState(0);
  const [order, setOrder] = useState([{ id: 1, adults: 1, children: 0 }]);

  const orderHandler = (id, value) => {
    order.map(data => {
      if (data.hasOwnProperty(id)) {
        console.log(data.id);
      }
      console.log(id, value);
    });

    setOrder();
  };
  return (
    <Container>
      <Wrapper>
        {rooms.map((room, i) => (
          <Form
            key={room.id}
            room={room}
            checked={i <= counter}
            onCheck={() => setCounter(i)}
            order={order}
            orderHandler={orderHandler}
          />
        ))}
      </Wrapper>
      <Button>Submit</Button>
    </Container>
  );
};

export default Main;

const Container = styled.div`
  box-sizing: border-box;
  display: grid;
  grid-template-rows: repeat(auto-fill, minmax(500px, 90% 10%));
  height: 50vh;
  width: 100%;
  padding: 20px;
`;

const Button = styled.button`
  height: 5vh;
  width: 6vw;
`;

const Wrapper = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(200px, 5fr));
  grid-gap: 20px;
  height: 150px;
  width: 100%;
`;
