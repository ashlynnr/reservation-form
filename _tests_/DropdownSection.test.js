import React from "react";
import { shallow } from "enzyme";
import rooms from "../utils/data";
import DropdownSection from "../components/DropdownSection";
import { handleChange } from "../components/DropdownSection";

describe("Dropdown Options", () => {
  // matches snapshot

  const defaultOrder = [{ id: 1, adults: 1, children: 0 }];
  const wrapper = shallow(
    <DropdownSection
      adults={[1, 2]}
      children={[1, 2, 3]}
      checked={true}
      id={1}
      orderHandler={jest.fn}
      order={defaultOrder}
    />
  );

  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("data to be accurate", () => {
    expect(rooms.length).toEqual(4);
  });

  it("Dropdowns to render", () => {
    const dropdown = wrapper.find("DropdownOptions").at(1);
    expect(dropdown).toBeDefined();
  });
});
