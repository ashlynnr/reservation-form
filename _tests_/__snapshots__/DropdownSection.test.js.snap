// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Options matches snapshot 1`] = `
<DropdownSection__Container>
  <DropdownOptions
    ages="(18+)"
    checked={true}
    currentValue={1}
    guestType="adults"
    handleChange={[Function]}
    options={
      Array [
        1,
        2,
      ]
    }
    title="Adults"
  />
  <DropdownOptions
    ages="(0-17)"
    checked={true}
    currentValue={0}
    guestType="children"
    handleChange={[Function]}
    options={
      Array [
        1,
        2,
        3,
      ]
    }
    title="Children"
  />
</DropdownSection__Container>
`;
